---
title: "Team Vibe Crew analysis"
output: html_notebook
---

# Load the required libraries
```{r setup}
library(Seurat)
library(ggplot2)
library(scCustomize)
library(readr)
library(pheatmap)
library(matrixStats)
library(spdep)
library(geojsonR)
library(SingleCellExperiment)
library(scuttle)
```

# Prepare the Xenium seurat object
```{r Load, eval=FALSE}
start <- Sys.time()

data_dir <- "/project/shared/spatial_data_camp/datasets/DATASET3/XENIUM_5K_LYMPH_NODE"

data <- ReadXenium(data_dir, outs = "matrix", type=c("centroids", "segmentations"))

end <- Sys.time()
```

```{r Benchmark, eval=FALSE}
paste0("Time to read in the data: ", round(end - start, 2), " mins")
```

```{r Metadata, eval=FALSE}
cell_meta_data <- read.csv(file.path(data_dir, "cells.csv.gz"))
annot <- read_csv("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Xenium_Prime_Human_Lymph_Node_Reactive_FFPE_cell_types.csv")

cell_meta_data <- merge(cell_meta_data, annot, by = "cell_id")
remove(annot)

head(cell_meta_data)
```

```{r, eval=FALSE}
seurat <- CreateSeuratObject(counts = data$matrix[["Gene Expression"]],
                                 assay = "XENIUM",
                                 meta.data = cell_meta_data)

seurat[["Negative.Control.Codeword"]] <- CreateAssayObject(counts = data$matrix[["Negative Control Codeword"]])
seurat[["Negative.Control.Probe"]] <- CreateAssayObject(counts = data$matrix[["Negative Control Probe"]])
seurat[["Unassigned.Codeword"]] <- CreateAssayObject(counts = data$matrix[["Unassigned Codeword"]])

coords <- CreateFOV(coords = list(centroids = CreateCentroids(data$centroids), 
                                  segmentation = CreateSegmentation(data$segmentations)),
                    type = c("segmentation", "centroids"),
                    assay = "XENIUM")
seurat[["LN"]] <- coords
```

```{r Delete unused objects and purge, eval=FALSE}
remove(list = list(data, coords, annot, cell_meta_data, start, end))
gc()
```

```{r Save/load pre-computed object}
#saveRDS(seurat, file = "/project/shared/spatial_data_camp/HACKATHON/vibe_crew/01_xenium_with_annot_and_fov_and_neg_probes.RDS")
seurat <- readRDS(file = "/project/shared/spatial_data_camp/HACKATHON/vibe_crew/01_xenium_with_annot_and_fov_and_neg_probes.RDS")
```

```{r Rename celltype column}
seurat$Celltype <- seurat$group
seurat$group <- NULL
```


# Subset on the Xenium explorer outline

```{r Read barcodes info}
barcodes <- read_csv("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/LN_whole_cells_stats.csv", skip = 2)
colnames(barcodes) <- c("cell_id", "annotation", "no_transcripts", "area")
```


```{r Visualise filter}
seurat$OUTLINE_FILTER <- seurat$cell_id %in% barcodes$cell_id
ImageDimPlot(seurat, group.by="OUTLINE_FILTER")
```

```{r Subset, warning=FALSE}
seurat <- subset(seurat, OUTLINE_FILTER)
ImageDimPlot(seurat, group.by="OUTLINE_FILTER")
```

```{r}
transcripts_counts_upper_outlier <- isOutlier(Sce_object$nCount_XENIUM, nmads = 2.5, share.mads = F,
                                type="higher", batch=Sce_object$Celltype)

transcripts_counts_lower_outlier <- isOutlier(Sce_object$nCount_XENIUM, nmads = 1.5, share.mads = F,
                                type="lower", batch=Sce_object$Celltype)
```


# Subset on the QC metrics

```{r Select filtering method}
method <- "MSD" #Select quantiles or MSD
```


## Transcript number
```{r}
if(method == "quantiles"){
  seurat$TRANSCRIPT_FILTER <- seurat$nCount_XENIUM >= 15
  ImageDimPlot(seurat, group.by="TRANSCRIPT_FILTER")
}
```

## Cell size (min and max)
```{r}
if(method =="quantiles"){
  seurat[["SIZE_FILTER_SMALL"]] <- seurat$cell_area > quantile(seurat$cell_area, .1)
  seurat[["SIZE_FILTER_LARGE"]] <- seurat$cell_area < quantile(seurat$cell_area, .99)
  ImageDimPlot(seurat, group.by="SIZE_FILTER_SMALL")
  ImageDimPlot(seurat, group.by="SIZE_FILTER_LARGE")
}
```

## Negative probes
```{r}
seurat[['PROBE_FILTER']] <- seurat$nCount_Unassigned.Codeword == 0 &
                       seurat$nCount_Negative.Control.Codeword == 0 &
                       seurat$nCount_Negative.Control.Probe == 0
```

## Per celltype filtering

```{r}
#Convert to SCE object
Sce_object <- as.SingleCellExperiment(seurat)

cells_celltype <- list()

for (i in unique(seurat$Celltype)) {
  cells_celltype[[i]] <- Sce_object[,rownames(as.data.frame(colData(Sce_object)) %>% dplyr::filter(Celltype == i))]
}

#Count cutoffs
upper_cut_off_counts_celltype <- list()
lower_cut_off_counts_celltype <- list()
figures <- list()

transcripts_counts_upper_outlier <- isOutlier(Sce_object$nCount_XENIUM, nmads = 2.5, share.mads = F,
                                type="higher", batch=Sce_object$Celltype)

transcripts_counts_lower_outlier <- isOutlier(Sce_object$nCount_XENIUM, nmads = 1.5, share.mads = F,
                                type="lower", batch=Sce_object$Celltype)

for (i in unique(seurat$Celltype)) {
  
  index <- match(i,unique(seurat$Celltype))
  
  upper_cut_off_counts_celltype[[i]] <- attr(transcripts_counts_upper_outlier, "thresholds")['higher', i]
  lower_cut_off_counts_celltype[[i]] <- attr(transcripts_counts_lower_outlier, "thresholds")['lower', i]
  
  data <- data.frame(x=cells_celltype[[i]]$nCount_XENIUM, 
                     Identity = cells_celltype[[i]]$Celltype)
  
  figures[[i]] <- ggplot(data,
                        aes(x = x, fill = as.factor(Identity))) + 
    geom_density(alpha = 0.5) +
    geom_vline(xintercept = lower_cut_off_counts_celltype[[i]], colour="grey", linetype = "longdash") +
    geom_vline(xintercept = upper_cut_off_counts_celltype[[i]], colour="grey", linetype = "longdash") +
    labs(x = expression('log'[10]*'(Library Size)'), title = "Total count density", fill = "Ident") + 
    theme_classic(base_size = 14)
  
} 

combined_plot_reads <- ggpubr::ggarrange(plotlist = figures, ncol = 2, nrow = 2, align = 'h')

# Get the number of unique subpopulations
num_subpopulations <- length(unique(seurat$Celltype))

# Calculate the number of chunks of 4 subpopulations
num_chunks <- ceiling(num_subpopulations / 4)

# Loop over the chunks
for (i in seq_len(num_chunks)) {
  # Calculate the start and end subpopulation for this chunk
  start_subpopulation <- (i - 1) * 4 + 1
  end_subpopulation <- min(i * 4, num_subpopulations)
  
  print(combined_plot_reads[[as.character(i)]])
  
}
#Cell size cutoffs
higher_cell_cut_off_size_celltype <- list()
lower_cell_cut_off_size_celltype <- list()
figures <- list()

cell_size_upper_outlier <- isOutlier(Sce_object$cell_area, nmads = 2.5, share.mads = F,
                                type="higher", batch=Sce_object$Celltype)

cell_size_lower_outlier <- isOutlier(Sce_object$cell_area, nmads = 1.5, share.mads = F,
                                type="lower", batch=Sce_object$Celltype)

for (i in unique(seurat$Celltype)) {
  
  index <- match(i,unique(seurat$Celltype))
  
  higher_cell_cut_off_size_celltype[[i]] <- attr(cell_size_upper_outlier, "thresholds")['higher', i]
  lower_cell_cut_off_size_celltype[[i]] <- attr(cell_size_lower_outlier, "thresholds")['lower', i]
  
  data <- data.frame(x=cells_celltype[[i]]$cell_area, 
                     Identity = cells_celltype[[i]]$Celltype)
  
  figures[[i]] <- ggplot(data,
                        aes(x = x, fill = as.factor(Identity))) + 
    geom_density(alpha = 0.5) +
    geom_vline(xintercept = higher_cell_cut_off_size_celltype[[i]], colour="grey", linetype = "longdash") +
    geom_vline(xintercept = lower_cell_cut_off_size_celltype[[i]], colour="grey", linetype = "longdash") +
    labs(x = expression('Cell Size'), title = "Total size density", fill = "Ident") + 
    theme_classic(base_size = 14)
  
} 

combined_plot_reads <- ggpubr::ggarrange(plotlist = figures, ncol = 2, nrow = 2, align = 'h')

# Get the number of unique subpopulations
num_subpopulations <- length(unique(seurat$Celltype))

# Calculate the number of chunks of 4 subpopulations
num_chunks <- ceiling(num_subpopulations / 4)

# Loop over the chunks
for (i in seq_len(num_chunks)) {
  # Calculate the start and end subpopulation for this chunk
  start_subpopulation <- (i - 1) * 4 + 1
  end_subpopulation <- min(i * 4, num_subpopulations)
  
  print(combined_plot_reads[[as.character(i)]])
  
}

#Cell Filtering

count.drop <- list()
size.drop <- list()

SCE_objects <- list()
SCE_objects.filtered <- list()

for (i in unique(Sce_object$Celltype)) {
  
  SCE_objects[[i]] <- Sce_object[,colnames(cells_celltype[[i]])]
  
}

for (i in unique(Sce_object$Celltype)) {

  count.drop[[i]] <- grep('TRUE', cells_celltype[[i]]$nCount_XENIUM >= lower_cut_off_counts_celltype[[i]] & cells_celltype[[i]]$nCount_XENIUM <= upper_cut_off_counts_celltype[[i]])
  size.drop[[i]] <- grep('TRUE', cells_celltype[[i]]$cell_area >= lower_cell_cut_off_size_celltype[[i]] & cells_celltype[[i]]$cell_area <= higher_cell_cut_off_size_celltype[[i]])
  
}

for (i in unique(Sce_object$Celltype)) {
  
  SCE_objects.filtered[[i]] <- SCE_objects[[i]][,intersect(count.drop[[i]], size.drop[[i]])]
  
}

cells_to_keep <- c()

for (i in unique(Sce_object$Celltype)) {
  
  index <- match(i, unique(Sce_object$Celltype))

  cells_to_keep <- c(cells_to_keep, colnames(SCE_objects.filtered[[i]]))
  
}

Sce_object_filtered <- Sce_object[,cells_to_keep]

# Export number of cells pre and post filtering

print(paste0("Total cells before quality filtering = ",dim(Sce_object)[2]))
print(paste0('Total cells before quality filtering, per sample:'))
table(colData(Sce_object)$Donor_subpopulation)

print(paste0("Total cells remaining after quality filtering = ",dim(Sce_object_filtered)[2]))
print(paste0('Total cells after quality filtering, per sample:'))
table(colData(Sce_object_filtered)$Celltype)

df_number_of_cells <- data.frame(row.names = c('Before_filtering', 'After_filtering'))

for (i in unique(Sce_object$Celltype)) {
  
  index <- match(i, unique(Sce_object$Celltype))

  df_number_of_cells['Before_filtering', i] <- ncol(SCE_objects[[i]])
  df_number_of_cells['After_filtering', i] <- ncol(SCE_objects.filtered[[i]])
  
}

# Export qc metrics pre and post filtering

df_qc <- data.frame()

for (i in unique(Sce_object$Celltype)) {
  
  index <- match(i, unique(Sce_object$Celltype))

  df_qc[i, 'Before_filtering__medianCount'] <- summary(colData(SCE_objects[[i]])$nCount_XENIUM)[[3]]
  df_qc[i, 'After_filtering__medianCount'] <- summary(colData(SCE_objects.filtered[[i]])$nCount_XENIUM)[[3]]
  df_qc[i, 'Before_filtering__medianSize'] <- summary(colData(SCE_objects[[i]])$nCount_XENIUM)[[3]]
  df_qc[i, 'After_filtering__medianSize'] <- summary(colData(SCE_objects.filtered[[i]])$nCount_XENIUM)[[3]]
  
}
```

```{r Save the result}
msd_barcodes <- colnames(Sce_object_filtered)
write_csv(data.frame("cell_id" = msd_barcodes), file = "msd_barcodes.csv")
```

```{r Clean-up}
remove(Sce_object)
remove(Sce_object_filtered)
remove(SCE_objects)
remove(SCE_objects.filtered)
```

```{r Filter the object}
if(method == "quantiles"){
  seurat$COMBINED_FILTER <- seurat$PROBE_FILTER & seurat$SIZE_FILTER_SMALL & seurat$SIZE_FILTER_LARGE & seurat$TRANSCRIPT_FILTER
} else if (method == "MSD"){
  seurat$COMBINED_FILTER <- seurat$cell_id %in% msd_barcodes
}

ImageDimPlot(seurat, group.by="COMBINED_FILTER")
```

```{r, warning=FALSE}
seurat <- subset(seurat, COMBINED_FILTER)
saveRDS(seurat, "/project/shared/spatial_data_camp/HACKATHON/vibe_crew/02_xenium_after_QC_filtering.RDS")
```

# Normalise the counts and calculate PC/UMAP

```{r, eval=FALSE}
seurat <- readRDS("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/02_xenium_after_QC_filtering.RDS")
```

```{r Normalise and PCA}
seurat <- SCTransform(seurat, assay = "XENIUM", clip.range = c(-10, 10))
seurat <- RunPCA(seurat)
```

```{r}
ElbowPlot(seurat, 50)
```

# Subset to a smaller object (just part of the lymph node)

```{r, reading in the filtered object, eval=FALSE}
seurat_small <- readRDS("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/02_xenium_after_QC_filtering.RDS")
```

```{r Read fragment barcodes info}
barcodes <- read_csv("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/LN_subset_cells_stats.csv", skip = 2)
colnames(barcodes) <- c("cell_id", "annotation", "no_transcripts", "area")
head(barcodes)
```

```{Subset to a fragment}
seurat_small$SMALL_FRAGMENT_FILTER <- seurat_small$cell_id %in% intersect(seurat_small$cell_id,barcodes$cell_id)
subset(seurat_small, group.by="SMALL_FRAGMENT_FILTER")
ImageDimPlot(seurat_small, group.by = "SMALL_FRAGMENT_FILTER")
```

```{r Normalise and PCA seurat small}
seurat_small <- SCTransform(seurat_small, assay = "XENIUM", clip.range = c(-10, 10))
seurat_small <- RunPCA(seurat_small)
```

```{r}
ElbowPlot(seurat_small, 50)
```

```{r}

```

# CellChat

```{r}
ptm = Sys.time()

library(CellChat)
library(patchwork)
options(stringsAsFactors = FALSE)
```

```{r}

png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/Phenotype/', 'Whole_slide_ALL_cells.png'), width = 12, height = 10, res = 300, units = 'in', type="cairo")
  
  plot <- ImageDimPlot(seurat, size = 1, 
          group.by='Celltype') + scale_fill_manual(values = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'), na.value = '#E6E7E9') + 
      theme(plot.title = element_text(hjust=0.5, vjust = 0.5),
          plot.subtitle = element_text(hjust=0.5, vjust = 0.5), legend.position = 'none') +
    ggtitle(label = "All", subtitle = sprintf('n = %s', ncol(seurat)))
  
  print(plot)
  
  dev.off()
```

```{r}
seurat_small <- readRDS("/project/shared/spatial_data_camp/HACKATHON/vibe_crew/04_xenium_small_after_QC_filtering_norm_PCA.RDS")
```

```{r}
seurat_small$Celltype_v2 <- seurat_small$Celltype %>% make.names

for (i in unique(seurat_small$Celltype)) {
  
  cells_highlight = WhichCells(seurat_small, expression = Celltype %in% i)
  
  name = unique(seurat_small@meta.data[cells_highlight,'Celltype_v2'])
  
  png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/Phenotype/', name, '.png'), width = 8, height = 10, res = 300, units = 'in', type="cairo")
  
  plot <- ImageDimPlot(seurat_small, 
          group.by='Celltype', cells = cells_highlight, size = 2) + scale_fill_manual(values = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'), na.value = '#E6E7E9') + 
      theme(plot.title = element_text(hjust=0.5, vjust = 0.5),
          plot.subtitle = element_text(hjust=0.5, vjust = 0.5),
          legend.position = 'none') +
    ggtitle(label = i, subtitle = sprintf('n = %s', length(cells_highlight)))
  
  print(plot)
  
  dev.off()
  
}

#

png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/Phenotype/', 'ALL_cells.png'), width = 12, height = 10, res = 300, units = 'in', type="cairo")
  
  plot <- ImageDimPlot(seurat_small, size = 1, 
          group.by='Celltype') + scale_fill_manual(values = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'), na.value = '#E6E7E9') + 
      theme(plot.title = element_text(hjust=0.5, vjust = 0.5),
          plot.subtitle = element_text(hjust=0.5, vjust = 0.5)) +
    ggtitle(label = "All", subtitle = sprintf('n = %s', ncol(seurat_small)))
  
  print(plot)
  
  dev.off()
```

```{r}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/Phenotype/', 'CD40_cells.png'), width = 12, height = 10, res = 300, units = 'in', type="cairo")
  
  plot <- ImageDimPlot(seurat_small, size = 2, 
          group.by='Celltype') + scale_fill_manual(values = c('Hematopoietic Stem Cell' = '#E6E7E9',
                    'Mast Cell' = '#E6E7E9',
                    'Neutrophil' = '#E6E7E9',
                    "Non Classical Monocyte" = '#E6E7E9',
                    'Intermediate Monocyte' = '#E6E7E9',
                    'Classical Monocyte' = '#E6E7E9',
                    "Macrophage" = '#E6E7E9',
                    'Erythrocyte' = '#E6E7E9',
                    'Plasma Cell' = '#E6E7E9',
                    'T Cell' = '#E6E7E9',
                    "CD4+ αβ Memory T Cell" = '#E6E7E9',
                    "Effector CD8+ αβ T Cell" = '#0000ff',
                    "Regulatory T Cell" = '#E6E7E9',
                    "Effector CD4+ αβ T Cell" = '#0000ff',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#0000ff',
                    "CD8+ αβ Memory T Cell" = '#E6E7E9',
                    "Stromal Cell" = '#E6E7E9',
                    'Mature Conventional Dendritic Cell' = '#E6E7E9',
                    'CD1c+ Myeloid Dendritic Cell' = '#E6E7E9',
                    'CD141+ Myeloid Dendritic Cell' = '#E6E7E9',
                    'Mature NK T Cell' = '#E6E7E9',
                    'Type I NK T Cell' = '#E6E7E9',
                    'Plasmacytoid Dendritic Cell' = '#E6E7E9',
                    'B Cell' = '#c1121f',
                    "Naive B Cell" = '#c1121f',
                    "Memory B Cell" = '#c1121f',
                    "Innate Lymphoid Cell" = '#E6E7E9',
                    "Endothelial Cell" = '#E6E7E9'), na.value = '#E6E7E9') + 
      theme(plot.title = element_text(hjust=0.5, vjust = 0.5),
          plot.subtitle = element_text(hjust=0.5, vjust = 0.5)) +
    ggtitle(label = "All", subtitle = sprintf('n = %s', ncol(seurat_small)))
  
  print(plot)
  
  dev.off()
```

```{r}
data.input = Seurat::GetAssayData(seurat_small, slot = "data", assay = "SCT")
seurat_small$orig.ident <- 'Dataset'
meta = seurat_small@meta.data
meta$samples <- factor(meta$orig.ident)

spatial.locs = Seurat::GetTissueCoordinates(seurat_small, scale = NULL, cols = c("imagerow", "imagecol")) 
rownames(spatial.locs) <- spatial.locs$cell
spatial.locs$cell <- NULL

spatial.factors = data.frame(samples = meta[,'samples'])
spatial.factors$ratio = 1 # The Xenium cell feature matrix and transcripts result coordinates are in microns.
spatial.factors$tol = 5 # Tolerance = half the size of a typical human cell 

cellchat <- createCellChat(object = data.input, meta = meta, group.by = "Celltype_v2",
                           datatype = "spatial", coordinates = spatial.locs, spatial.factors = spatial.factors)
```

```{r}
cellchat
```


```{r}
CellChatDB <- CellChatDB.human
showDatabaseCategory(CellChatDB)
```

```{r}
CellChatDB.use <- subsetDB(CellChatDB)
cellchat@DB <- CellChatDB.use
```

```{r}
cellchat <- subsetData(cellchat)
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat, variable.both = F)
 
execution.time = Sys.time() - ptm
print(as.numeric(execution.time, units = "secs"))
```

```{r}
cellchat <- computeCommunProb(cellchat, type = "truncatedMean", trim = 0.1, population.size = T,
                              distance.use = TRUE, interaction.range = 250, scale.distance = 0.1,
                              contact.dependent = TRUE, contact.range = 50)
```


```{r}
cellchat <- filterCommunication(cellchat, min.cells = 10)
```

```{r}
cellchat <- computeCommunProbPathway(cellchat)
```

```{r}
cellchat <- aggregateNet(cellchat)

execution.time = Sys.time() - ptm
print(as.numeric(execution.time, units = "secs"))
```

```{r}
dir.create(path = '/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', showWarnings = F)
```

```{r}
ptm = Sys.time()

groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$weight, vertex.weight = rowSums(cellchat@net$weight), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```

```{r}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Number_of_LR_chord_plot.png'), width = 10, height = 6, res = 300, units = 'in', type="cairo")

netVisual_circle(cellchat@net$count, vertex.weight = rowSums(cellchat@net$count), weight.scale = T, label.edge= F, title.name = "Number of interactions", color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))

dev.off()
```

```{r}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Strength_of_LR_chord_plot.png'), width = 10, height = 6, res = 300, units = 'in', type="cairo")

netVisual_circle(cellchat@net$weight, vertex.weight = rowSums(cellchat@net$weight), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength", color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))

dev.off()
```

```{r}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Number_of_LR.png'), width = 10, height = 8, res = 300, units = 'in', type="cairo")

netVisual_heatmap(cellchat, measure = "count", color.heatmap = "OrRd", width = 16, height = 12,
                                  color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))


dev.off()
```

```{r}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Strength_of_LR.png'), width = 10, height = 8, res = 300, units = 'in', type="cairo")

netVisual_heatmap(cellchat, measure = "weight", color.heatmap = "OrRd", width = 16, height = 12,
                                  color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))


dev.off()
```

```{r}
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred
```

```{r fig.height = 12, fig.width = 12, fig.align = "center"}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Outgoing_signalling.png'), width = 12, height = 12, res = 300, units = 'in', type="cairo")

netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing", width = 16, height = 12, 
                                  color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))

dev.off()
```

```{r fig.height = 12, fig.width = 12, fig.align = "center"}
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', 'Incoming_signalling.png'), width = 12, height = 12, res = 300, units = 'in', type="cairo")

netAnalysis_signalingRole_heatmap(cellchat, pattern = "incoming", width = 16, height = 12,
                                  color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                    'Mast Cell' = '#FB7A13',
                    'Neutrophil' = '#06d6a0',
                    "Non Classical Monocyte" = '#00a8e8',
                    'Intermediate Monocyte' = '#007ea7',
                    'Classical Monocyte' = '#003459',
                    "Macrophage" = '#5a189a',
                    'Erythrocyte' = '#c1121f',
                    'Plasma Cell' = '#C1B793',
                    'T Cell' = '#7D676D',
                    "CD4+ αβ Memory T Cell" = '#d5bdaf',
                    "Effector CD8+ αβ T Cell" = '#a9927d',
                    "Regulatory T Cell" = '#e0afa0',
                    "Effector CD4+ αβ T Cell" = '#c0c0c0',
                    "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                    "CD8+ αβ Memory T Cell" = '#9381ff',
                    "Stromal Cell" = '#fca311',
                    'Mature Conventional Dendritic Cell' = '#99582a',
                    'CD1c+ Myeloid Dendritic Cell' = '#432818',
                    'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                    'Mature NK T Cell' = '#a4c3b2',
                    'Type I NK T Cell' = '#6b9080',
                    'Plasmacytoid Dendritic Cell' = '#61B4EC',
                    'B Cell' = '#B4D5A1',
                    "Naive B Cell" = '#90be6d',
                    "Memory B Cell" = '#a7c957',
                    "Innate Lymphoid Cell" = '#ff8fab',
                    "Endothelial Cell" = '#540b0e'))

dev.off()
```

```{r}
for (i in cellchat@netP$pathways) {

  pathways.show <- i
  
  png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', i, '_LRs_chord_plot.png'), width = 10, height = 8, res = 300, units = 'in', type="cairo")
  
  netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle",
                                    color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                      'Mast Cell' = '#FB7A13',
                      'Neutrophil' = '#06d6a0',
                      "Non Classical Monocyte" = '#00a8e8',
                      'Intermediate Monocyte' = '#007ea7',
                      'Classical Monocyte' = '#003459',
                      "Macrophage" = '#5a189a',
                      'Erythrocyte' = '#c1121f',
                      'Plasma Cell' = '#C1B793',
                      'T Cell' = '#7D676D',
                      "CD4+ αβ Memory T Cell" = '#d5bdaf',
                      "Effector CD8+ αβ T Cell" = '#a9927d',
                      "Regulatory T Cell" = '#e0afa0',
                      "Effector CD4+ αβ T Cell" = '#c0c0c0',
                      "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                      "CD8+ αβ Memory T Cell" = '#9381ff',
                      "Stromal Cell" = '#fca311',
                      'Mature Conventional Dendritic Cell' = '#99582a',
                      'CD1c+ Myeloid Dendritic Cell' = '#432818',
                      'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                      'Mature NK T Cell' = '#a4c3b2',
                      'Type I NK T Cell' = '#6b9080',
                      'Plasmacytoid Dendritic Cell' = '#61B4EC',
                      'B Cell' = '#B4D5A1',
                      "Naive B Cell" = '#90be6d',
                      "Memory B Cell" = '#a7c957',
                      "Innate Lymphoid Cell" = '#ff8fab',
                      "Endothelial Cell" = '#540b0e'))
  
  dev.off()
  
  #
  
  png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', i, '_LRs_chord_plot_v2.png'), width = 14, height = 12, res = 300, units = 'in', type="cairo")
  
  group.cellType <- levels(cellchat@idents)
  names(group.cellType) <- levels(cellchat@idents)
  netVisual_chord_cell(cellchat, signaling = pathways.show, group = group.cellType, title.name = paste0(pathways.show, " signaling network",
                                    color.use = c('Hematopoietic Stem Cell' = '#82a0bc',
                      'Mast Cell' = '#FB7A13',
                      'Neutrophil' = '#06d6a0',
                      "Non Classical Monocyte" = '#00a8e8',
                      'Intermediate Monocyte' = '#007ea7',
                      'Classical Monocyte' = '#003459',
                      "Macrophage" = '#5a189a',
                      'Erythrocyte' = '#c1121f',
                      'Plasma Cell' = '#C1B793',
                      'T Cell' = '#7D676D',
                      "CD4+ αβ Memory T Cell" = '#d5bdaf',
                      "Effector CD8+ αβ T Cell" = '#a9927d',
                      "Regulatory T Cell" = '#e0afa0',
                      "Effector CD4+ αβ T Cell" = '#c0c0c0',
                      "Naive Thymus Derived CD4+ αβ T Cell" = '#efcfe3',
                      "CD8+ αβ Memory T Cell" = '#9381ff',
                      "Stromal Cell" = '#fca311',
                      'Mature Conventional Dendritic Cell' = '#99582a',
                      'CD1c+ Myeloid Dendritic Cell' = '#432818',
                      'CD141+ Myeloid Dendritic Cell' = '#bb9457',
                      'Mature NK T Cell' = '#a4c3b2',
                      'Type I NK T Cell' = '#6b9080',
                      'Plasmacytoid Dendritic Cell' = '#61B4EC',
                      'B Cell' = '#B4D5A1',
                      "Naive B Cell" = '#90be6d',
                      "Memory B Cell" = '#a7c957',
                      "Innate Lymphoid Cell" = '#ff8fab',
                      "Endothelial Cell" = '#540b0e')))
                      
  dev.off()
  
  #
  
  png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', i, '_LRs_sender_received_heatmap.png'), width = 18, height = 3, res = 300, units = 'in', type="cairo")
  
  netAnalysis_signalingRole_network(cellchat, signaling = pathways.show, width = 18, height = 3, font.size = 10)
  
  dev.off()
  
}
```
```{r}
library(ggdark)

for (lr in cellchat@net$LRs) {
  
png(filename = paste0('/project/shared/spatial_data_camp/HACKATHON/vibe_crew/Figures/CellChat/', lr, '_LR_projection.png'), width = 12, height = 12, res = 300, units = 'in', type="cairo")

plot <- spatialFeaturePlot(cellchat, pairLR.use = lr, point.size = 3, do.binary = TRUE, cutoff = 0.05, enriched.only = F, color.heatmap = "Reds", direction = 1) + dark_theme_void()

print(plot)

dev.off()
  
}
```





